PhysicsTools/PatAlgos/plugins/PATTauProducer.cc

```diff
@@ -309,8 +309,9 @@ void PATTauProducer::produce(edm::Event & iEvent, const edm::EventSetup & iSetup
     if ( addTauID_ ) {
       std::string missingDiscriminators;
       std::vector<pat::Tau::IdPair> ids(tauIDSrcs_.size());
+      auto const& tausDeref = *tausRef;
       for ( size_t i = 0; i < tauIDSrcs_.size(); ++i ) {
-	if ( typeid(*tausRef) == typeid(reco::PFTau) ) {
+	if ( typeid(tausDeref) == typeid(reco::PFTau) ) {
```

Comment:
since you're modifying this, wouldn't it be cleaner to do `auto const & tauTypeId = typeid(*tausRef)` at the beginning, and not re-call typeid()?

Reply:
It is the call `typeid(*tausRef)` which the compiler is complaining about. Basically, `*tausRef` calls `operator*()` which is a complicated function which changes the state of `tausRef`. Normally, `typeid` is compile time resolvable statement, in this case it is not. clang appears to be warning that the `typeid(*tausRef)` is not as trivial as most people might expect for a call to `typeid`.

Reply:
As for not recalling `typeid`, I bet with my modification, the compiler is already optimizes it such that `typeid` is called only once.

Reply:
So I checked using gcc.godbolt.org using gcc 7.3 and `--std=c++17 -O2` with both just calling `typeid` each time versus `auto const& tid = typeid(...)`. The machine code generated is identical.